### PR TITLE
Semi-working kfctl go binary

### LINKED ISSUE #2472: kfctl: support deploy with basic auth
User should be able to deploy cluster with go binary while provide username/password to later access their kubeflow service

### LINKED ISSUE #2491: kfctl go binary should have an option to skip initi project
The kfctl go binary should have an option to skip initializing the GCP project.

### LINKED ISSUE #2514: add additional config files to kfctl for iap, basic_auth, default components
in order to accommodate the different configuration possibilities we should allow kfctl to load config from 
- https://github.com/kubeflow/kubeflow/blob/master/bootstrap/config/kfctl_iap.yaml
- https://github.com/kubeflow/kubeflow/blob/master/bootstrap/config/kfctl_basic_auth.yaml
- https://github.com/kubeflow/kubeflow/blob/master/bootstrap/config/kfctl_defaults.yaml

this will allow a different set of packages, components and parameters than standard as well as iap,  basic_auth

### PR DESCRIPTION
**Pull request by gabrielwen**
Works related to this PR:
- Fixes #2491
- Fixes #2472
- Fixes #2514
- Making #2370 ~80% finished.

Known WIPs:
- #2586 
- #2585 
- #2507 
- #2535
- #2391
- #2604
- When creating resources, some of resources has error message `roles.rbac.authorization.k8s.io [resource] is forbidden: attempt to grant extra privileges`
- Error message `Resource creation for metricscollector-template is failed at 1 attempt: ConfigMap "" is invalid: metadata.name: Required value: name or generateName is required`
- #2605 pod `iap` doesn't seem to be working - Having error message like #953

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/2587)

### CODE REVIEW

**Comment by lluunn** [comment]
could you add in PR description which part of gcp is done and what's remaining?

**Comment by gabrielwen** [comment]
> could you add in PR description which part of gcp is done and what's remaining?

Added some, those are the WIPs that I know of.  I believe there are more as we dig into it, tho.

File: bootstrap/config/types.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package config
+
+import ()
```
Comment (lluunn) [inline bootstrap/config/types.go]: remove?

File: bootstrap/config/types.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+type NameValue struct {
```
Comment (lluunn) [inline bootstrap/config/types.go]: add comment on the new types

File: bootstrap/pkg/apis/apps/group.go [hunk 3]
```diff
@@ -37,9 +45,12 @@ const (
 	DefaultPlatform  = "ksonnet"
 	// TODO: find the latest tag dynamically
 	DefaultVersion  = "master"
-	DefaultGitRepo  = "https://github.com/kubeflow/kubeflow/tarball"
+	DefaultGitRepo  = "https://github.com/gabrielwen/kubeflow/tarball"
```
Comment (lluunn) [inline bootstrap/pkg/apis/apps/group.go]: revert?

File: bootstrap/cmd/kfctl/cmd/apply.go [hunk 4]
```diff
@@ -69,4 +69,20 @@ func init() {
 		log.Errorf("couldn't set flag --%v: %v", string(kftypes.VERBOSE), bindErr)
 		return
 	}
+	applyCmd.Flags().String(string(kftypes.OAUTH_ID), "",
+		"OAuth Client ID, GCP only. Required if not ENV CLIENT_ID is not set. "+
```
Comment (lluunn) [inline bootstrap/cmd/kfctl/cmd/apply.go]: if not --> if. same below.

File: bootstrap/pkg/utils/k8utils.go [hunk 5]
```diff
@@ -38,10 +44,152 @@ import (
...
+}
+
+func getResource(mapping *meta.RESTMapping, config *rest.Config, group string,
+	version string, namespace string, name string) error {
+	restClient, err := getRESTClient(config, group, version)
+	if err != nil {
+		return fmt.Errorf("getResource error: %v", err)
+	}
+
+	_, err = restClient.
```
Comment (lluunn) [inline bootstrap/pkg/utils/k8utils.go]: get should return the response?

File: bootstrap/config/types.go (hunk 1 above)
Comment (gabrielwen) [reply bootstrap/config/types.go]: removed.

File: bootstrap/pkg/apis/apps/group.go (hunk 3 above)
Comment (gabrielwen) [reply bootstrap/pkg/apis/apps/group.go]: reverted.

File: bootstrap/cmd/kfctl/cmd/apply.go (hunk 4 above)
Comment (gabrielwen) [reply bootstrap/cmd/kfctl/cmd/apply.go]: updated.

File: bootstrap/config/types.go (hunk 2 above)
Comment (gabrielwen) [reply bootstrap/config/types.go]: added.

**Comment by lluunn** [comment]
if we plan to shell out the kubectl, do we still need the k8sutil thing?
One reason I added that was because we don't want to shell out in bootstrapper until we isolate different runs in containers.
@kunmingg

File: bootstrap/pkg/utils/k8utils.go (hunk 5 above)
Comment (gabrielwen) [reply bootstrap/pkg/utils/k8utils.go]: currently we don't use the response.  added a TODO in case we need to do it.

**Comment by gabrielwen** [comment]
> if we plan to shell out the kubectl, do we still need the k8sutil thing?
> One reason I added that was because we don't want to shell out in bootstrapper until we isolate different runs in containers.
> @kunmingg

I think we definitely should try to avoid shell out especially in bootstrapper.  It'll be very hard to poll the status and it's gonna stuck there for minutes - not ideal for server loading.

**Comment by jlewi** [comment]
> When creating resources, some of resources has error message `roles.rbac.authorization.k8s.io [resource] is forbidden: attempt to grant extra privileges`

This is likely because you need to explicitly grant the current user cluster-admin privileges
Like we do here:
https://github.com/kubeflow/kubeflow/blob/544bf699370c59a019c7218fd06d72d42e377221/scripts/gke/util.sh#L179

**Comment by gabrielwen** [comment]
> > When creating resources, some of resources has error message `roles.rbac.authorization.k8s.io [resource] is forbidden: attempt to grant extra privileges`
> 
> This is likely because you need to explicitly grant the current user cluster-admin privileges
> Like we do here:
> [kubeflow/scripts/gke/util.sh](https://github.com/kubeflow/kubeflow/blob/544bf699370c59a019c7218fd06d72d42e377221/scripts/gke/util.sh#L179)
> 
> Line 179 in [544bf69](/kubeflow/kubeflow/commit/544bf699370c59a019c7218fd06d72d42e377221)
> 
>  kubectl create clusterrolebinding default-admin --clusterrole=cluster-admin --user=${EMAIL}

agree - need to use impersonate as admin for these requests.

**Comment by gabrielwen** [comment]
will address to other comments tomorrow :)

**Comment by gabrielwen** [comment]
> Do we need to do impersonate? I don't think we use impersonation with kfctl.sh. What is click to deploy currently doing?
> 
> > _[Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/2587#-:-LZu5DvABVe1XhOFI5RL:b-5dfhpd)_ status: 13 of 15 files reviewed, 21 unresolved discussions (waiting on @ellis-bigelow, @gabrielwen, @jlewi, and @lluunn)

We do: https://github.com/kubeflow/kubeflow/blob/master/scripts/gke/util.sh#L197

File: bootstrap/config/kfctl_iap.yaml [hunk 6]
```diff
@@ -3,23 +3,46 @@
...
+  - application
+  - modeldb
+  - gcp
+components:
+  - pytorch-operator
   - ambassador
+  - openvino
+  - jupyter
+  - notebook-controller
+  - jupyter-web-app
```
Comment (kunmingg) [inline bootstrap/config/kfctl_iap.yaml]: openvino is not a default component, shouldn't be part of default config.
 We can add it on the fly via go logic.

File: bootstrap/config/kfctl_iap.yaml [hunk 7]
```diff
@@ -3,23 +3,46 @@
...
   - centraldashboard
   - tf-job-operator
+  - tensorboard
+  - metacontroller
+  - profiles
+  - notebooks
```
Comment (kunmingg) [inline bootstrap/config/kfctl_iap.yaml]: We are still working on ```notebooks``` and ```notebook-controller``` components.
We can skip them in this PR and add later once work on that side is done.

**Comment by kunmingg** [comment]
@gabrielwen 
Instead of impersonate you can bind ```default-admin``` to user before do apply.
click to deploy app is "Out of Cluster" as well.

File: bootstrap/config/kfctl_iap.yaml (hunk 6 above)
Comment (gabrielwen) [reply bootstrap/config/kfctl_iap.yaml]: removed.

File: bootstrap/config/kfctl_iap.yaml (hunk 7 above)
Comment (gabrielwen) [reply bootstrap/config/kfctl_iap.yaml]: removed.

File: bootstrap/config/kfctl_iap.yaml [hunk 8]
```diff
@@ -30,13 +53,22 @@ parameters:
       value: ipName
     - name: hostname
       # TODO change value on the fly: replace with user-provide parameters. This need to be fully qualified domain name to use with ingress.
-      value: <deployName>.endpoints.<Project>.cloud.goog
+      value: <deployName>.endpoints.<project>.cloud.goog
+compParams:
```
[the remark this answers is not in the document]
Comment (gabrielwen) [reply bootstrap/config/kfctl_iap.yaml]: `protoParams` is called with `ks init`, while `compParams` is called with `ks param set`.  Explaned in bootstrap/config/types.go

**Comment by gabrielwen** [comment]
> @gabrielwen
> Instead of impersonate you can bind `default-admin` to user before do apply.
> click to deploy app is "Out of Cluster" as well.

mm, it's weird that:
- I already bind "default-admin" to user.
- If I use `kubectl apply`, this kind of error doesn't show up.

**Comment by gabrielwen** [comment]
> > @gabrielwen
> > Instead of impersonate you can bind `default-admin` to user before do apply.
> > click to deploy app is "Out of Cluster" as well.
> 
> mm, it's weird that:
> 
> * I already bind "default-admin" to user.
> * If I use `kubectl apply`, this kind of error doesn't show up.

**Comment by gabrielwen** [comment]
> @gabrielwen
> Instead of impersonate you can bind `default-admin` to user before do apply.
> click to deploy app is "Out of Cluster" as well.

Also, from [here](https://github.com/kubeflow/kubeflow/blob/76230712d80805251f2ac54d66b0f02838de1c79/bootstrap/cmd/bootstrap/app/ksServer.go#L431) I think it's using In Cluster config.

**Pull request merged**